Solutions/Marain.Workflow.Api.EngineHost.Client/Microsoft/Extensions/DependencyInjection/WorkflowEngineClientServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,62 @@
+﻿// <copyright file="WorkflowEngineClientServiceCollectionExtensions.cs" company="Endjin">
+// Copyright (c) Endjin. All rights reserved.
+// </copyright>
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    using System;
+    using System.Linq;
+    using Corvus.Extensions.Json;
+    using Corvus.Identity.ManagedServiceIdentity.ClientAuthentication;
+    using Marain.Workflows.Client;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Rest;
+
+    /// <summary>
+    /// DI initialization for clients of the workflow engine service.
+    /// </summary>
+    public static class WorkflowEngineClientServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds the workflow engine client to a service collection.
+        /// </summary>
+        /// <param name="services">The service collection.</param>
+        /// <param name="resourceIdForMsiAuthentication">
+        /// The resource id to use when obtaining an authentication token representing the
+        /// hosting service's identity. Pass null to run without authentication.
+        /// </param>
+        /// <returns>The modified service collection.</returns>
+        public static IServiceCollection AddMarainWorkflowEngineClient(
+            this IServiceCollection services,
+            Func<IServiceProvider, MarainWorkflowEngineOptions> getOptions,
+            string resourceIdForMsiAuthentication = null)
```

Comment:
Although this has been refactored to support callback-based passing of options, that's sort of half-baked, because the resource ID still needs to be passed in during service collection initialization time. And since this will normally be supplied via configuration, we won't actually have that.

If you look at https://github.com/marain-dotnet/Marain.Tenancy/blob/e9a8dd9ea5465fa02f6f66082be122ba5f43350c/Solutions/Marain.Tenancy.Client/Marain/Tenancy/Client/TenancyClientServiceCollectionExtensions.cs you'll see that does it by putting _all_ the settings into the `TenancyClientOptions` type, including the `ResourceIdForMsiAuthentication`. (In fact, it also avoids even needing a callback to obtain the settings—it just pulls the `TenancyClientOptions` directly out of DI instead of making you provide a callback, although it's debatable whether that's necessarily an improvement.)

So I think you'll need to move this `resourceIdForMsiAuthentication` into the options type here too, as otherwise you're not really getting the benefit of that `getOptions` callback. (The point of that is to deal with the fact that Azure Functions don't actually let us get hold of our `IConfiguration` until _after_ we've finished configuring services.)